### PR TITLE
PoW mining more like PoS implementation

### PR DESCRIPTION
**Pull request by codingupastorm**
If the PoW miner throws a consensus error exception during block assembly, mining will be permanently halted until the node is restarted. This PR changes this so that mining will continue even if an exception is encountered - the implementation is mostly taken from the PosMinting class. Feel like these two classes could implement the same interface in the future and then MiningFeature could be a bit cleaner too.

This PR will be helpful for the upcoming smart contract nodes which currently use PoW.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Miner/MiningFeature.cs [hunk 1]
```diff
@@ -158,7 +162,7 @@ public override void Initialize()
         /// <inheritdoc />
         public override void Dispose()
         {
-            this.powLoop?.Dispose();
```
Comment (codingupastorm) [inline src/Stratis.Bitcoin.Features.Miner/MiningFeature.cs]: @noescape00 this is exactly what was happening anyway though?

Not that it's 'correct', just that the functionality hasn't changed.

File: src/Stratis.Bitcoin.Features.Miner/PowMining.cs [hunk 2]
```diff
@@ -83,22 +82,49 @@ public class PowMining : IPowMining
...
-            repeatEvery: TimeSpans.RunOnce,
+            repeatEvery: TimeSpans.Second,
             startAfter: TimeSpans.TenSeconds);
+        }
 
-            return this.mining;
+        ///<inheritdoc/>
+        public void StopMining()
+        {
+            this.mining.Dispose();
```
[the remark this answers is not in the document]
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Miner/PowMining.cs]: Ah yes the loop uses `this.nodeLifetime.ApplicationStopping` 
You have to do like as POS and create a linked cancellation 
```
this.stakeCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(new[] { this.nodeLifetime.ApplicationStopping });
```

**Comment by dangershony** [comment]
Looks good to me just create a new cancellation token for the miner and lets push

**Comment by codingupastorm** [comment]
Some thoughts about the Pow/Pos interfaces, having looked at them more deeply now:

- `GenerateBlocks/GenerateBlocksAsync` shouldn't be public without having a way to inject the token cancellation source. If you call `GenerateBlocksAsync` on PosMinting right now then `stakeCancellationTokenSource.Token` is going to give a NullReferenceException. The only way to run this method correctly is via `Stake()` anyway. As such there are no unit tests for this public method on PosMinting right now. 

-This method is tested for PowMining however. So to allow the tests to pass with the new cancellation source I am initialising the cancellation token in the constructor.

I'd really like to work on this area in the future, though it could be a while before I can take a look. I think we could:
-Make the different "miners" somewhat consistent and share an interface
-Have both the actual mining method (GenerateBlocks) and the loop creation method (Mine/Stake) public and work standalone, and be tested

**Comment by fassadlr** [comment]
@codingupastorm Hey Jordan, looks good to me as well. Can you confirm the following before we merge:

1> All unit tests pass.
2> All integration tests pass.
3> You have synced the node in testnet and mainnet from IBD.
4> You have tested that mining and staking still works and can be stopped.

**Comment by bokobza** [comment]
Looks good to me too.
I don't think doing `You have synced the node in testnet and mainnet from IBD.` is necessary for this.

**Comment by codingupastorm** [comment]
Have done 1,2,4, all good!

**Pull request merged**